frontend/packages/shared/src/hooks/mutations/useUpsertTextResourcesMutation.ts

```diff
@@ -12,31 +12,48 @@ export interface UpsertTextResourcesMutationArgs {
   textResources: ITextResource[];
 }
 
+export type UseUpsertTextResourceMutationResult = UseMutationResult<
+  ITextResourcesWithLanguage,
+  DefaultError,
+  UpsertTextResourcesMutationArgs
+>;
+
 export const useUpsertTextResourcesMutation = (
   org: string,
   app: string,
-): UseMutationResult<UpsertTextResourcesMutationArgs> => {
+): UseUpsertTextResourceMutationResult => {
   const previewConnection = usePreviewConnection();
   const { upsertTextResources } = useServicesContext();
   const queryClient = useQueryClient();
-  return useMutation({
-    mutationFn: ({ language, textResources }: UpsertTextResourcesMutationArgs) =>
+  const queryKey = [QueryKey.TextResources, org, app];
+  return useMutation<ITextResourcesWithLanguage, DefaultError, UpsertTextResourcesMutationArgs>({
+    mutationFn: ({
+      language,
+      textResources,
+    }: UpsertTextResourcesMutationArgs): Promise<ITextResourcesWithLanguage> =>
       upsertTextResources(
         org,
         app,
         language,
         TextResourceUtils.fromArray(textResources).toObject(),
-      ).then(() => ({ language, textResources })),
-    onSuccess: async ({ language, textResources }: UpsertTextResourcesMutationArgs) => {
+      ),
+    onMutate: ({ language, textResources }: UpsertTextResourcesMutationArgs): void => {
+      queryClient.setQueryData<ITextResources>(
+        queryKey,
+        (oldTexts: ITextResources): ITextResources =>
+          setTextResourcesForLanguage(oldTexts, language, textResources),
+      );
+    },
+    onError: () => queryClient.invalidateQueries({ queryKey }),
+    onSuccess: async (response: ITextResourcesWithLanguage): Promise<void> => {
       if (previewConnection && previewConnection.state === 'Connected') {
         await previewConnection.send('sendMessage', 'reload-layouts').catch(function (err) {
           return console.error(err.toString());
         });
       }
       queryClient.setQueryData(
         [QueryKey.TextResources, org, app],
```

Comment:
Can we use the local queryKey variable here too?
```suggestion
        queryKey,
```

Reply:
Fixed